### PR TITLE
Avoid learning the same job multiple times

### PR DESCRIPTION
**Pull request by pietroalbini**
The current learning implementation was designed with Travis CI in mind, where a single check was sent when the job ended. Because of that, when a successful build was detected all the successful job outputs were learned. This assumption doesn't hold anymore for GitHub Actions though. 

GHA sends a check for every job in the matrix, and that causes RLA to learn a successful job every time one of the following jobs in the same build finishes, learning it multiple times. For example, if a build contains 60 jobs, the first job to finish will be learned 60 times.

This fixes the problem by remembering which job IDs we learned before, and skipping the learning process if the ID was previously seen. Along with that the code to remember IDs was refactored.

r? @Mark-Simulacrum

### CODE REVIEW

File: src/bin/server/worker.rs [hunk 1]
```diff
@@ -341,3 +332,70 @@ impl Worker {
...
+impl<T: Clone + Eq + Hash> RecentlySeen<T> {
+    fn new(size: usize) -> Self {
+        Self {
+            size,
+            lookup: HashSet::new(),
+            removal: VecDeque::new(),
+        }
+    }
+
+    fn check(&self, key: &T) -> bool {
```
Comment (Mark-Simulacrum) [inline src/bin/server/worker.rs:356]: Can we call this something less ambiguous? Maybe contains or recently_witnessed?

File: src/bin/server/worker.rs [hunk 2]
```diff
@@ -341,3 +332,70 @@ impl Worker {
...
+    lookup: HashSet<T>,
+    removal: VecDeque<T>,
+}
+
+impl<T: Clone + Eq + Hash> RecentlySeen<T> {
+    fn new(size: usize) -> Self {
+        Self {
+            size,
+            lookup: HashSet::new(),
+            removal: VecDeque::new(),
```
Comment (Mark-Simulacrum) [inline src/bin/server/worker.rs:352]: nit: it probably makes sense to immediately with_capacity these with size, though at our scale it certainly doesn't matter

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me, primarily only care about the method rename though

**Comment by pietroalbini** [comment]
Addressed all the feedback.

**Review by Mark-Simulacrum — APPROVED** [review]

**Pull request merged**